server/chat-plugins/the-studio.ts

```diff
@@ -0,0 +1,613 @@
+	],
+
+	denysuggestion(target, room, user) {
+		const targetRoom = Rooms.search('thestudio') || room;
+		Recs.checkRoom(targetRoom);
+		this.room = targetRoom;
+		this.checkCan('mute', null, targetRoom!);
+		const [submitter, artist, title] = target.split('|').map(x => x.trim());
+		if (!(submitter && artist && title)) return this.parse(`/help approvesuggestion`);
+
+		Recs.denySuggestion(submitter, artist, title);
+
+		this.privateModAction(`${user.name} denied a suggested recommendation from ${submitter} for '${title}' by ${artist}.`);
+		this.modlog(`RECOMMENDATION`, null, `deny: '${toID(title)}' by ${toID(artist)} from ${submitter}`);
+	},
+	denysuggestionhelp: [
+		`/denysuggestion submitter | artist | strong title - Deny a submitted song recommendation. Requires: % @ * # &`,
+	],
+
+	rec: 'recommendation',
+	searchrec: 'recommendation',
+	viewrec: 'recommendation',
+	searchrecommendation: 'recommendation',
+	viewrecommendation: 'recommendation',
+	randrec: 'recommendation',
+	randomrecommendation: 'recommendation',
+	async recommendation(target, room, user) {
+		if (!recommendations.saved.length) {
+			throw new Chat.ErrorMessage(`There are no recommendations saved.`);
+		}
+		const targetRoom = Rooms.search('thestudio') || room;
+		Recs.checkRoom(targetRoom);
+		this.room = targetRoom;
+		this.runBroadcast();
+		if (!target) {
+			return this.sendReply(`|html|${await Recs.render(Recs.getRandomRecommendation())}`);
+		}
+		const matches: Recommendation[] = [];
+		if (target) {
+			target = target.slice(0, 300);
+			const args = target.split(',');
+			for (const rec of recommendations.saved) {
+				if (!args.every(x => rec.tags.map(toID).includes(toID(x)))) continue;
+				matches.push(rec);
+			}
+		}
+		if (!matches.length) {
+			throw new Chat.ErrorMessage(`No matches found.`);
+		}
+		const sample = Utils.shuffle(matches)[0];
+		this.sendReply(`|html|${await Recs.render(sample)}`);
+	},
+	recommendationhelp: [
+		`/recommendation [key1, key2, key3, ...] - Displays a random recommendation that matches all keys, if one exists.`,
+		`If no arguments are provided, a random recommendation is shown.`,
+	],
+
+	likerec: 'likerecommendation',
+	likerecommendation(target, room, user, connection) {
+		const targetRoom = Rooms.search('thestudio') || room;
+		Recs.checkRoom(targetRoom);
+		this.room = targetRoom;
+		const [artist, title] = target.split('|').map(x => x.trim());
+		if (!(artist && title)) return this.parse(`/help likerecommendation`);
+		Recs.likeRecommendation(artist, title, user);
+		this.sendReply(`You liked '${title}' by ${artist}.`);
+	},
+	likerecommendationhelp: [
+		`/likerecommendation artist | title - Upvotes a recommendation for the provided artist and title.`,
+	],
+
+	viewrecs: 'viewrecommendations',
+	viewrecommendations(target, room, user) {
+		const targetRoom = Rooms.search('thestudio') || room;
+		Recs.checkRoom(targetRoom);
```

Comment:
Why not just do this in one function, passing commandcontext, like youtube does?

Reply:
bc it doesnt work as cleanly for setting `this.room` for parsing commands from a page